test/watcher_test.js

```diff
@@ -139,6 +162,65 @@ describe('Watcher', function() {
       expect(builderBuild).to.have.been.called;
     });
 
+    it('on change, rebuild is invoked and cancel is invoked from the build', function() {
+      const pluginsCalled = {};
+
+      const waitForPlugin1 = defer();
+      const waitForPlugin2 = defer();
+
+      class Plugin1 extends Plugin {
+        constructor(inputNodes, options) {
+          super(inputNodes || [], options);
+        }
+
+        build() {
+          pluginsCalled['plugin1'] = pluginsCalled['plugin1'] ? pluginsCalled['plugin1'] + 1 : 1;
+
+          return waitForPlugin1.promise;
+        }
+      }
+
+      class Plugin2 extends Plugin {
+        constructor(inputNodes, options) {
+          super(inputNodes || [], options);
+        }
+
+        build() {
+          pluginsCalled['plugin2'] = pluginsCalled['plugin2'] ? pluginsCalled['plugin2'] + 1 : 1;
+
+          return waitForPlugin2.promise;
+        }
+      }
+
+      const plugin1 = new Plugin1();
+      const plugin2 = new Plugin2();
```

Comment:
Passing arguments to `PluginX` to avoid needing to customize `constructor` (and therefore simplifying the test a bit):

```suggestion
      const plugin2 = new Plugin2([]);
```